libraries/AC_WPNav/AC_WPNav.cpp

```diff
@@ -1040,27 +1040,26 @@ float AC_WPNav::get_slow_down_speed(float dist_from_dest_cm, float accel_cmss)
 /// wp_speed_update - calculates how to handle speed change requests
 void AC_WPNav::wp_speed_update(float dt)
 {
-    // calculate speed change for steady-state or speeding up
-    if (_wp_desired_speed_cms > _wp_speed_cms) {
-        _wp_speed_cms += _wp_desired_speed_cms;
-        if (_wp_speed_cms > _wp_desired_speed_cms) {
-            _wp_speed_cms = _wp_desired_speed_cms;
+    // calculate speed change otherwise return 
+    if (_wp_desired_speed_cms > _wp_current_speed_cms) {
+        // speed up is requested so increase speed within limit set by WPNAV_ACCEL
+        _wp_current_speed_cms += _wp_accel_cmss * dt;
+        if (_wp_current_speed_cms > _wp_desired_speed_cms) {
+            _wp_current_speed_cms = _wp_desired_speed_cms;
         }
-        //update position controller speed
-        _pos_control.set_max_speed_xy(_wp_speed_cms);
-    
-        // flag that wp leash must be recalculated
-        _flags.recalc_wp_leash = true;
-    } else if (_wp_desired_speed_cms < _wp_speed_cms) {
+    } else if (_wp_desired_speed_cms < _wp_current_speed_cms) {
         // slow down is requested so reduce speed within limit set by WPNAV_ACCEL
-        _wp_speed_cms -= _wp_accel_cmss * dt;
-        if (_wp_speed_cms < _wp_desired_speed_cms) {
-            _wp_speed_cms = _wp_desired_speed_cms;
+        _wp_current_speed_cms -= _wp_accel_cmss * dt;
+        if (_wp_current_speed_cms < _wp_desired_speed_cms) {
+            _wp_current_speed_cms = _wp_desired_speed_cms;
         }
-        //update position controller speed
-        _pos_control.set_max_speed_xy(_wp_speed_cms);
-    
-        // flag that wp leash must be recalculated
-        _flags.recalc_wp_leash = true;
+    } else {
+        return;
```

Comment:
maybe let's try the early return using the is_equal() as suggested by Peter?

Reply:
Will do.  I should have that completed tonight.